---
title: "Human GC content"
author: "Laura Biggins"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: kate
---

```{r setup, echo=FALSE, cache=FALSE, message = FALSE, warning = FALSE}
library(knitr)
library(rmdformats)
## Global options
options(max.print="75")
display_code <- FALSE
knitr::opts_knit$set(global.par = TRUE, width=75)
knitr::opts_chunk$set(
                echo    = display_code,
                prompt  = FALSE,
                tidy    = TRUE,
                comment = NA,
                message = FALSE,
                warning = FALSE)
                #cache=TRUE)

library(rmdformats)
library(runGOA)
library(magrittr)
library(dplyr)
library(ggplot2)
library(beanplot)

par(mgp = c(2, 0.5, 0))
par(mar = par()$mar * 0.7)
#load_all()
min_appearances <- 20
```

```{r DT_function}
dt_setup <-  function(data, n_rows = 10, lineHeight="80%") {
  DT::datatable(data,
                rownames = FALSE,
                options = list(dom = 't', 
                              pageLength = n_rows)
  ) %>%
  DT::formatStyle(0, target = 'row', lineHeight = lineHeight)
}

dark_b_blue <- "#374772" 
light_b_blue <- "#3cc1f2"
```

### Generate lists of human genes by chromosome.

Use a gene info file derived from a gtf file from Ensembl containing all the genes
in the human genome.

```{r import_genfo_file}
# The gene info file contains the number of GCs of all the genes in the human genome. 
# The processing of it is described in the processing_gtf file.
load("../data/genfo.rda")
load("../data/all_go_categories_human.rda")

all_GO_annotated_genes <- unique(as.vector(unlist(all_go_categories_human)))

chr_order <- c(1:22, "MT", "X", "Y")

genfo_all_genes <- genfo %>%
    mutate(chromosome = factor(chromosome, levels = chr_order))

genfo <- genfo %>%
    filter(gene_name %in% all_GO_annotated_genes) %>%
    mutate(chromosome = factor(chromosome, levels = chr_order))
```


## Plot the number of genes on each chromosome in human genome

### Including every gene in the gtf file

```{r, fig.width = 10}
genfo_all_genes %>%
    count(chromosome) %>%
    ggplot(aes(x= chromosome, y = n)) +
        geom_col(fill = light_b_blue, colour = dark_b_blue, lwd = 1) +
        ylab("number of genes on chromosome")
```


### Only genes with GO annotations

This is what we'll use for generating the lists of genes.

```{r}
genfo %>%
    count(chromosome) %>%
    ggplot(aes(x= chromosome, y = n)) +
        geom_col(fill = light_b_blue, colour = dark_b_blue, lwd = 1) +
        ylab("number of genes on chromosome")

```

```{r}
dt_setup(count(genfo, chromosome))
```

```{r}
get_genes_for_chr <- function(chr){
    
    filter(genfo, chromosome == chr) %>%
        sample_n(size = 200) %>%
        pull(gene_name)
}

```

<br><br>

## Generating the biased gene lists and running them through functional enrichment analysis

1. Filter the set of genes for those on specified chromosome.
2. 200 genes are then randomly selected from the filtered set. 
3. These are run through a functional enrichment analysis and the results are recorded.

The steps are repeated 100 times for each chromosome.


```{r GO_analysis, eval = FALSE}
# This code is run once to get the GO analysis results- it takes a while to run
# it 100 times for each bias category.
load("../data/all_background_genes_human.rda")

# set up the summary files so we can populate them as we get the GO analysis results
go_category_counts <- tibble(GO_category = names(all_go_categories_human))

summary_stats <- tibble(
                      run_number         = rep(1:100, times = length(chr_order)), 
                      bias_category      = rep(chr_order, each = 100), 
                      sig_number_results = 0)

summary_stats$sig_number_results %<>% as.integer


for (i in 1:length(chr_order)){
    
    bias_category <- chr_order[i]
   
    go_category_counts <- mutate(go_category_counts, occurrences = 0)

    for(j in 1:100) {

        genes <- get_genes_for_chr(bias_category)

        go_results <- overrep_test(all_go_categories_human, 
                                   query_genes = genes, 
                                   all_background_genes_human, 
                                   unique_genes_in_categories = all_GO_annotated_genes,
                                   min_genes_in_category = 1, 
                                   max_genes_in_category = 500000,
                                   ease = FALSE)
        
        if(length(go_results) > 0){
           
            summary_stats <- summary_stats %>%
                mutate(sig_number_results = if_else(
                    bias_category == !!bias_category & run_number == j, 
                    nrow(go_results), 
                    sig_number_results
                ))
        }  
        go_category_counts <- go_category_counts %>%
            mutate(occurrences = if_else(GO_category %in% rownames(go_results), 
                                         occurrences + 1, 
                                         occurrences)
                   ) 
    }

    # I couldn't get tidyverse to work properly with mutating column names and adding to them, so we'll rename afterwards
    go_category_counts <- rename(go_category_counts, !!bias_category := occurrences)

}

human_chr_biased  <- go_category_counts
human_chr_summary <- summary_stats

usethis::use_data(human_chr_biased, overwrite = TRUE)
usethis::use_data(human_chr_summary, overwrite = TRUE)
```

<div class = "row">
<div class = "col-md-6">
```{r}
load("../data/human_chr_biased.rda")
load("../data/human_chr_summary.rda")

category_counts <- human_chr_summary %>%
    filter(sig_number_results > 0) %>%
    group_by(bias_category) %>%
    count(bias_category)

dt_setup(category_counts)
```
</div>
<div class = "col-md-6">
<br><br>
The table shows the number of functional enrichment analyses that returned 1 or 
more significantly enriched categories from the 100 that were run. 
</div> 
</div>
</div> 
<br>

## Identified categories

The number of times that a functional category was identified from the 100 times
that the analysis was run for each bias category.
The table contains every functional category that appeared once or more in any 
of the bias categories.

```{r}
all_identified_categories <- human_chr_biased %>%
  filter_at(vars(-GO_category), any_vars(. > 0))

dt_setup(all_identified_categories)  
```

In total, `r nrow(all_identified_categories)` functional categories appeared 
at least once.


## Filter Categories

We don't want categories that only appeared a handful of times so we'll use an 
arbitrary value of `r min_appearances` as a threshold.

```{r}
filtered_categories <- human_chr_biased %>%
  filter_at(vars(-GO_category), any_vars(. > min_appearances))

dt_setup(filtered_categories)  
```


```{r}
suspects_for_violin <- human_chr_biased %>%
  tidyr::gather(-GO_category, key = "chr", value = "occurrences") %>%
  filter(occurrences > min_appearances) 
```

```{r fig.height = 4, fig.width = 5}
total_counts <- suspects_for_violin %>%
  count(bias_category)

suspects_for_violin  %>%
    ggplot(aes(x = bias_category, y = occurrences)) +
    geom_violin(fill = light_b_blue, colour = dark_b_blue, lwd = 0.8) +
    xlab("") +
    ylab("number of times category was identified") +
    theme_bw() +
    theme(axis.text.x = element_text(size = 12)) +
    geom_dotplot(binaxis='y', stackdir='center', dotsize=0.1, colour = dark_b_blue) +
    geom_label(data = total_counts, colour = dark_b_blue, fill = "lightgrey", aes(label = paste0("n=",n), x = 1:nrow(total_counts), y = 100)) 
```


```{r} 
# eval=FALSE, echo = FALSE}
# write out file with unix line endings
for (i in 1:length(suspects)) {
  
  filename <- paste0("../data/", names(suspects)[i], ".txt")
  output_file <- file(filename, "wb")
  
  write.table(file = output_file, 
              x = suspects[[i]], 
              row.names = FALSE,
              col.names = FALSE, 
              quote = FALSE)
  
  close(output_file)
}
```
